### PR TITLE
First pass at using an Application class

### PR DESCRIPTION
**Pull request by afshin**
First pass at using an `Application` class to build a Phosphide app with its chrome being injected into an `IAppShell` instance by the `Application` instance instead of menus and command palette and other core plugins attempting to inject themselves. This allows for core plugin consumers to set their own language and their own application layout.

### CODE REVIEW

**Comment by sccolbert** [comment]
Why did you move the palette widget into the plugin file?

**Comment by afshin** [comment]
Because that's where `AppShell` is and I wanted them to have the same directory structure.

**Comment by sccolbert** [comment]
Appshell  (as of now) is not intended to be subclassed or reused. I thought we had talked about maybe making command palette public in the future (like side bar).

**Comment by afshin** [comment]
Ah, right. I had written it with the intention of it being subclassable, but it's not public. Good point. @dwillmer is gonna be pleased with me breaking his merge yet again! I'll put it in its own file and plugin will a simple file with the `register` function and a wrapper class that has the static `requires` and `create` methods.

**Comment by sccolbert** [comment]
Yep. Wrapper class is a good idea. That will let us work on the API of the widget to make it generally useful, but only expose a restricted API in the wrapper class. The commands should be defined by the wrapper class which call into public APIs of the widget.

**Comment by afshin** [comment]
@sccolbert I'm not sure why the wrapper class should do anything other than take care of dependencies. I made `CommandPalette` just be a simple widget, not a plugin, and the wrapper simply takes care of its dependencies and the wrapper's `create` returns an instance of an `ICommandPalette` (namely, an actual instance of a `CommandPalette`).

Do you think the wrapper needs to be more than this: https://github.com/afshin/phosphide/blob/743a5b6021f2d4e66644a5cf2078881ad6b0744c/src/commandpalette/plugin.ts

**Comment by blink1073** [comment]
I agree with @afshin, we're not at 1.0 yet, I say make the `CommandPallette` stand on its own and we can change the API as needed.

**Comment by dwillmer** [comment]
yep, agreed

**Comment by sccolbert** [comment]
Yes I think it should be different. The widget shouldn't define a token if its standalone. Also when you define a class like this, the create method should return an instance of the class. 

If you are creating an instance of a different class, all you need is a factory object literal. That is, you arent instantiating the wrapper class, which means it shouldn't be a class at all.

**Comment by sccolbert** [comment]
Actually, I guess it has to define the token since ICommandPalette extends widget

**Comment by sccolbert** [comment]
And since the palette depends on a command registry, theres not much point in ever moving it out of phosphide. So a factory object should be sufficient here.

**Comment by afshin** [comment]
Cool, I just switched to an `IFactory`.

File: src/commandpalette/plugin.ts [hunk 1]
```diff
@@ -49,27 +42,5 @@ import {
...
-    this._palette.title.text = 'Commands';
-  }
-
-  run(): void {
-    this._shell.addToLeftArea(this._palette, { rank: 40 });
-  }
-
-  private _shell: IAppShell;
-  private _palette: ICommandPalette;
+  container.register(ICommandPalette, paletteFactory);
```
Comment (blink1073) [inline src/commandpalette/plugin.ts]: You can just instantiate the factory here and avoid the separate variable: see https://github.com/blink1073/jupyter-js-plugins/blob/browser-factory/src/filebrowser/plugin.ts

Comment (afshin) [inline src/commandpalette/plugin.ts]: Cool. I guess if it's a singleton, having the `const` doesn't make any sense. I removed that.

**Comment by blink1073** [comment]
LGTM.

**Pull request merged**